safenode/src/network/cmd.rs

```diff
@@ -58,6 +77,38 @@ pub struct SwarmLocalState {
 impl SwarmDriver {
     pub(crate) fn handle_cmd(&mut self, cmd: SwarmCmd) -> Result<(), Error> {
         match cmd {
+            SwarmCmd::GetData { key, sender } => {
+                match self.swarm.behaviour_mut().kademlia.store_mut().get(&key) {
+                    Some(data) => {
+                        sender
+                            .send(GetChunk(Ok(Chunk::new(Bytes::from(data.value.clone())))))
+                            .map_err(|_| Error::InternalMsgChannelDropped)?;
+                    }
+                    None => {
+                        // empty response here just to test out
+                        sender
+                            .send(GetChunk(Err(ProtocolError::RecordNotFound)))
+                            .map_err(|_| Error::InternalMsgChannelDropped)?;
+
+                        warn!("No data found in kademlia store for key: {:?}", key);
+                    }
+                }
+            }
+            SwarmCmd::RegisterProvidedData { record } => {
+                let _ = self
+                    .swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .start_providing(record.key.clone())?;
+
+                // TODO: Does Kad automatically remove irrelevant records?
+                // Or do we need to manage this manually as we move about the network?
+                let _ = self
+                    .swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .put_record(record, libp2p::kad::Quorum::All)?;
```

Comment:
Maybe we could remove the provider stuff here and just roll with the PUT/GET flow for now. My understanding is that, to fetch data using the provider API, we should first make a call to `kad::get_providers(key)` and use then contact the peers (providers) for our data. But currently we’re just calling get_closest_peer(key) and then requesting the data from them directly. And the "provider" nodes are not serving any purpose.

Also, the PUT/GET flow using `.put_record()` is the one responsible for the actual "data" replication. So it might be better to just start with that and add in providers later.

Reply:
- `start_providing(record.key.clone())?` - produces `ProviderRecords` which are replicated among the `closest_peers(key)`.
- `.put_record(record, libp2p::kad::Quorum::All)?` - produces `Records` which are replicated among the `closest_peers(key)`